web/middlewares/secure.go

```diff
@@ -0,0 +1,112 @@
+package middlewares
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cozy/cozy-stack/pkg/instance"
+	"github.com/labstack/echo"
+)
+
+type (
+	XFrameOption string
+	CSPSource    int
+
+	// SecureConfig defines the config for Secure middleware.
+	SecureConfig struct {
+		HSTSMaxAge    time.Duration
+		CSPScriptSrc  []CSPSource
+		CSPFrameSrc   []CSPSource
+		CSPConnectSrc []CSPSource
+		XFrameOptions XFrameOption
+		XFrameAllowed string
+	}
+)
+
+const (
+	XFrameDeny       XFrameOption = "DENY"
+	XFrameSameOrigin              = "SAMEORIGIN"
+	XFrameAllowFrom               = "ALLOW-FROM"
+
+	CSPSrcSelf CSPSource = iota
+	CSPSrcParent
+	CSPSrcParentSubdomains
+	CSPSrcAny
+)
+
+func Secure(conf *SecureConfig) echo.MiddlewareFunc {
+	var hstsHeader string
+	if conf.HSTSMaxAge > 0 {
+		hstsHeader = fmt.Sprintf("max-age=%.f; includeSubdomains",
+			conf.HSTSMaxAge.Seconds())
+	}
+
+	var xFrameHeader string
+	switch conf.XFrameOptions {
+	case XFrameDeny:
+		xFrameHeader = string(XFrameDeny)
+	case XFrameSameOrigin:
+		xFrameHeader = string(XFrameSameOrigin)
+	case XFrameAllowFrom:
+		xFrameHeader = fmt.Sprintf("%s %s", XFrameAllowFrom, conf.XFrameAllowed)
+	}
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			hsts := true
+			if in := c.Get("instance"); in != nil && in.(*instance.Instance).Dev {
+				hsts = false
+			}
+			h := c.Response().Header()
+			if hsts && hstsHeader != "" {
+				h.Set(echo.HeaderStrictTransportSecurity, hstsHeader)
+			}
+			if xFrameHeader != "" {
+				h.Set(echo.HeaderXFrameOptions, xFrameHeader)
+			}
+			var cspHeader string
+			host := c.Request().Host
+			if len(conf.CSPScriptSrc) > 0 {
+				cspHeader += makeCSPHeader(host, "script-src", conf.CSPScriptSrc)
```

Comment:
I think we should allow inline-script because it seems to be very restrictive without that when I tested it. Maybe for us we can figured out (it would be great to not need them at all) but that could complicate the app development by the community. What's your opinion about that? We could also try for now without and see if we have troubles.